core/types.rs

```diff
@@ -387,39 +387,78 @@ pub struct OwnedRecord {
+const I16_LOW: i64 = -32768;
+const I16_HIGH: i64 = 32767;
+const I24_LOW: i64 = -8388608;
+const I24_HIGH: i64 = 8388607;
+const I32_LOW: i64 = -2147483648;
+const I32_HIGH: i64 = 2147483647;
+const I48_LOW: i64 = -140737488355328;
+const I48_HIGH: i64 = 140737488355327;
+
+// https://www.sqlite.org/fileformat.html#record_format
+const SERIAL_TYPE_INTEGER_ZERO: u64 = 0;
+const SERIAL_TYPE_I8: u64 = 1;
+const SERIAL_TYPE_I16: u64 = 2;
+const SERIAL_TYPE_I24: u64 = 3;
+const SERIAL_TYPE_I32: u64 = 4;
+const SERIAL_TYPE_I48: u64 = 5;
+const SERIAL_TYPE_I64: u64 = 6;
+const SERIAL_TYPE_F64: u64 = 7;
+
 impl OwnedRecord {
     pub fn new(values: Vec<OwnedValue>) -> Self {
         Self { values }
     }
 
     pub fn serialize(&self, buf: &mut Vec<u8>) {
         let initial_i = buf.len();
+        let mut serial_types = Vec::with_capacity(self.values.len());
 
+        // First pass: calculate serial types and store them
         for value in &self.values {
             let serial_type = match value {
-                OwnedValue::Null => 0,
-                OwnedValue::Integer(_) => 6, // for now let's only do i64
-                OwnedValue::Float(_) => 7,
+                OwnedValue::Null => SERIAL_TYPE_INTEGER_ZERO,
+                OwnedValue::Integer(i) => match i {
+                    i if *i >= I8_LOW && *i <= I8_HIGH => SERIAL_TYPE_I8,
+                    i if *i >= I16_LOW && *i <= I16_HIGH => SERIAL_TYPE_I16,
+                    i if *i >= I24_LOW && *i <= I24_HIGH => SERIAL_TYPE_I24,
+                    i if *i >= I32_LOW && *i <= I32_HIGH => SERIAL_TYPE_I32,
+                    i if *i >= I48_LOW && *i <= I48_HIGH => SERIAL_TYPE_I48,
+                    _ => SERIAL_TYPE_I64,
+                },
+                OwnedValue::Float(_) => SERIAL_TYPE_F64,
                 OwnedValue::Text(t) => (t.value.len() * 2 + 13) as u64,
                 OwnedValue::Blob(b) => (b.len() * 2 + 12) as u64,
                 // not serializable values
                 OwnedValue::Agg(_) => unreachable!(),
                 OwnedValue::Record(_) => unreachable!(),
             };
 
-            buf.resize(buf.len() + 9, 0); // Ensure space for varint
+            buf.resize(buf.len() + 9, 0); // Ensure space for varint (1-9 bytes in length)
             let len = buf.len();
             let n = write_varint(&mut buf[len - 9..], serial_type);
             buf.truncate(buf.len() - 9 + n); // Remove unused bytes
+
+            serial_types.push(serial_type);
         }
 
         let mut header_size = buf.len() - initial_i;
         // write content
-        for value in &self.values {
-            // TODO: make integers and floats with smaller serial types
+        for (value, &serial_type) in self.values.iter().zip(serial_types.iter()) {
             match value {
                 OwnedValue::Null => {}
-                OwnedValue::Integer(i) => buf.extend_from_slice(&i.to_be_bytes()),
+                OwnedValue::Integer(i) => match serial_type {
```

Comment:
is it better to just recalculate serial type here instead of allocating a vec above